### ISSUE TITLE
Module 5 Feedback

### ISSUE THREAD

**Issue by hfboyce** [UBC-MDS/introduction-machine-learning#27]
@mgelbart 

Welcome to episode 5 of intro to machine learning. 

On this week's episode we learn about preprocessing -> https://ml-learn.mds.ubc.ca/en/module5

The biggest change since my last push is this module should have 21 exercises. So if you don't have this when you click the link, clear your cookies, refresh galore and if still not changing message me and I'll see what's up. 

 Assignment 5 will be on it's way in a few days. 

Also HAPPY HALLOWEEN! 🎃 👻 🧚‍♀️ 🧛‍♀️

**Comment by mgelbart** [comment]
Here is what I have prior to today's meeting:

- 0.2: the pipeline LO is repeated twice. Also remove "preliminary" from it. 
- I think we should use the term "overfitting on the validation set" rather than optimization bias. That seems to be more standard, optimization bias seems to be a Mark thing. (Though I do think it's a good name)
- 1.3: I like to try using something they know (e.g. KNN) and see it throw an error, and then says this is because of the non-numeric features. Update: Oh I see, that is for Module 6 isn't it? OK. Let's show this in Module 6 then!
- 1.4: Not sure if this really matters, but if you only want the CV score you can use `cross_val_score` instead of `cross_validate`.
- 1.5: Not sure what went on in Module 4., but might it be less confusing to do `euclidean_distances(two_players)[1,0]` ? Also, the slide is cut off here, I can't see the result at the bottom. 
- 1.7: I think the last line would be clearer as `X_train_transformed = transformer.transform(X_train)`
- 1.8: the result is cut off at the bottom for me 
- 2.1: let's remove DummyClassifier - I think it's kind of a valid answer
- I'm a bit confused by 5.3. Discuss in meeting today. I also have a comment on the code. 
- 5.6: showing the error -> great
- 5.9: "constant" is also a valid approach for categorical variables I think
- 6.2: change "can we" to "should we"
- 8: I get

```
FileNotFoundError: [Errno 2] File data/bball_imp.csv does not exist: 'data/bball_imp.csv'
``` 

let me know when fixed and then I'll review 8.

- 9.3: cool. Who made this diagram? Though the RobustScaler doesn't showcase that well in this case I guess, given the lack of outliers in the raw data.
- I guess in hindsight it could have been find to just stick with one scaler here, but since you've already done the work it's fine. But this makes me think, for CPSC 330, maybe I should teach StandardScaler and RobustScaler instead of StandardScaler and MinMaxScaler...

**Comment by mgelbart** [comment]
Squeezed in a bit more pre-meeting, sorry for this being spread out like this.

- 9.6, 9.8: training score is not that meaningful in general - can we see the test scores?
- 10.3: it doesn't allow me to select the answer
- 12: per my earlier comment, it feels a bit redundant to do both here. They are more or less the same thing in most cases. 
- 13.5: I think it would be less confusing to show the transformation before `train_test_split`, rather than gluing them back together.
- 13.7: not sure if this is worth changing, but I mildly prefer the syntax:

```python
pipe = Pipeline([
        ("imputer", SimpleImputer(strategy="median")),
        ("scaler", StandardScaler()),
        ("reg", KNeighborsRegressor())
])
```

There is also `make_pipeline`, but I think this is probably better for when we get to the hyperparameter optimization.

-  13.8: we should show the code for these steps. let's split this into two slides. On the first, we show the pipeline `fit` side-by-side with the steps broken down. On the second slide, we do the same for `predict`

**Comment by mgelbart** [comment]
- 13.9:  The bottom of the diagram is cut off for me. 
- 17.2: Remove reference to deep learning
- 17.4: I think we should show this after we show them the code, rather than before.
- 17.6: This may be a bit overwhelming. I'd rather show them `GridSearchCV` with only one hyperparameter, outside of a Pipeline. Then move to multiple hyperparameters. And then move to the `__` stuff from Pipelines. 
- 17.7: `fit in training portion with the given C` -> `fit in training portion with the given C and gamma`
- 17.8: bottom is cut off for me 
- 17.9: just a delightful tidbit for you and @kvarada regarding `grid_search` vs. `grid_search.best_estimator_`: as stated in the slides, `predict` and `score` are the same. But `cross_val_score` is not the same, because `cross_val_score(grid_search, X, y)` calls `fit` on `grid_search`, which tries to do the whole hyperparameter optimization each fold! I learned that the hard way!
- 17.12: I wonder if we should omit these for these students. Curious to hear @kvarada 's thoughts. Instead, we should put in a range instead of a grid. Because `C` and `gamma` are continuous hypers, a realistic use case of randomized search would be to put in a min and max value, and let it choose uniformly at random in the range. The example we showed, where it picks randomly from a grid, is not that realistic for numeric hypers.
- 17.14: Change name to overfitting on validation set
- 17.15: consider replacing this slide with the figure from the top of my CPSC 330 lecture 6?
- 18: I think we should say "Exhaustive Grid Search (`GridSearchCV`)". I was thinking about the same thing when you had standardization/normalization MCQ earlier, might have been nice to say "Standardization (`StandardScaler`)". Or did you omit these on purpose? Also Randomize -> Randomized
- 18.2: Change to "Which one gives you fine-grained control over the amount of time spent searching?" Answer: Randomized
- Should we discuss implications of `n_jobs=-1` on our servers? Would it be better to omit this? Discuss in meeting?
- 20: I got the first MCQ wrong, whoops! But, do we mention the refitting anywhere in the slides? I think this merits its own dedicated slide.
- 21: same comments as for Ex 0

This all looks good @hfboyce , nice work 🙌

**Comment by hfboyce** [comment]
@mgelbart 

> 1.3: I like to try using something they know (e.g. KNN) and see it throw an error, and then says this is because of the non-numeric features. Update: Oh I see, that is for Module 6 isn't it? OK. Let's show this in Module 6 then!

We do show it throwing an error for imputation! 

> 1.7: I think the last line would be clearer as X_train_transformed = transformer.transform(X_train)

The current last line is 
```
transformer.fit_transform(X_train)
``` 
And that's because I added notes saying *We can also carry out fitting and transforming in one call using fit_transform, but we must be mindful to use it only on the train split and not on the test split.*

>  I'm a bit confused by 5.3. Discuss in meeting today. I also have a comment on the code.

Dropping the columns after talking. 

> 9.3: cool. Who made this diagram? Though the RobustScaler doesn't showcase that well in this case I guess, given the lack of outliers in the raw data.

Keeping in per discussion changed width. 

>  13.5: I think it would be less confusing to show the transformation before train_test_split, rather than gluing them back together.

I don't understand what you mean here. 

> 17.6: This may be a bit overwhelming. I'd rather show them GridSearchCV with only one hyperparameter, outside of a Pipeline. Then move to multiple hyperparameters. And then move to the __ stuff from Pipelines.

Can you take a look now and see if this is more up your alley? 

> 17.12: I wonder if we should omit these for these students. Curious to hear @kvarada 's thoughts. Instead, we should put in a range instead of a grid. Because C and gamma are continuous hypers, a realistic use case of randomized search would be to put in a min and max value, and let it choose uniformly at random in the range. The example we showed, where it picks randomly from a grid, is not that realistic for numeric hypers.

Need confirmation on where to go here. 

> 17.15: consider replacing this slide with the figure from the top of my CPSC 330 lecture 6?

I am looking [here](https://github.com/UBC-CS/cpsc330/blob/master/lectures/06_overfitting-validation-and-categoricals.ipynb) and do you mean the graph? 

>  Should we discuss implications of n_jobs=-1 on our servers? Would it be better to omit this? Discuss in meeting?

I add notes on this in 17 where we lightly say it's to run things in parallel and make things faster.

> 20: I got the first MCQ wrong, whoops! But, do we mention the refitting anywhere in the slides? I think this merits its own dedicated slide.

I think we do briefly when we say we are making a new model.

**Comment by hfboyce** [comment]
Oh Also I fixed Question 8!

**Comment by mgelbart** [comment]
> > 1.3: I like to try using something they know (eg. KNN) and see it throw an error, and then says this is because of the non-numeric features. Update: Oh I see, that is for Module 6 isn't it? OK. Let's show this in Module 6 then!
> 
> We do show it throwing an error for imputation!

Yeah I saw that later. Let's do that again in Module 6 with categorical variables.

> 
> > 1.7: I think the last line would be clearer as X_train_transformed = transformer.transform(X_train)
> 
> The current last line is
> 
> ```
> transformer.fit_transform(X_train)
> ```
> 
> And that's because I added notes saying _We can also carry out fitting and transforming in one call using fit_transform, but we must be mindful to use it only on the train split and not on the test split._

That is fair, but this is the training set here in the code, so that seems like a separate point? What I wanted to emphasize is that `fit_transform` returns the transformed data. If you only had `transformer.fit_transform(X_train)` then you might as well do `transformer.fit(X_train)` because the `_transform` part is being thrown away. So it feels like a confusing demonstration of how `fit_transform` is used. 

> 
> > I'm a bit confused by 5.3. Discuss in meeting today. I also have a comment on the code.
> 
> Dropping the columns after talking.
> 
> > 9.3: cool. Who made this diagram? Though the RobustScaler doesn't showcase that well in this case I guess, given the lack of outliers in the raw data.
> 
> Keeping in per discussion changed width.
> 
> > 13.5: I think it would be less confusing to show the transformation before train_test_split, rather than gluing them back together.
> 
> I don't understand what you mean here.

In the current version, we call `train_test_split` on an earlier slide. Then here we glue them back together and then scale to show that this is a bad idea. Instead, why not just take the original `df` from way back before `train_test_split` was ever called?

> 
> > 17.6: This may be a bit overwhelming. I'd rather show them GridSearchCV with only one hyperparameter, outside of a Pipeline. Then move to multiple hyperparameters. And then move to the __ stuff from Pipelines.
> 
> Can you take a look now and see if this is more up your alley?

Not sure if I'm seeing the new version or if the refresh failed. Can you send me a screenshot to confirm?

> 
> > 17.12: I wonder if we should omit these for these students. Curious to hear @kvarada 's thoughts. Instead, we should put in a range instead of a grid. Because C and gamma are continuous hypers, a realistic use case of randomized search would be to put in a min and max value, and let it choose uniformly at random in the range. The example we showed, where it picks randomly from a grid, is not that realistic for numeric hypers.
> 
> Need confirmation on where to go here.

Let's go with my suggestion, I don't think it's that contentious.

> 
> > 17.15: consider replacing this slide with the figure from the top of my CPSC 330 lecture 6?
> 
> I am looking [here](https://github.com/UBC-CS/cpsc330/blob/master/lectures/06_overfitting-validation-and-categoricals.ipynb) and do you mean the graph?

Yeah. But on reconsideration, I think what you did on 17.15 is quite nice. Maybe we can keep it in, but then add my graph on the next slide. And the transcript text for the graph is that is shows 1) cross-validation score is too optimistic vs. test score and 2) cross-validation might not pick the optimal hyperparameters.

> 
> > Should we discuss implications of n_jobs=-1 on our servers? Would it be better to omit this? Discuss in meeting?
> 
> I add notes on this in 17 where we lightly say it's to run things in parallel and make things faster.

OK. I'm just wondering if it'll actually be faster if there's tons of students using the same server. Or if the server might get upset by multiple people trying to use all 8 cores at once. But I guess it's fine. I wonder why `n_jobs=-1` isn't the default.

> 
> > 20: I got the first MCQ wrong, whoops! But, do we mention the refitting anywhere in the slides? I think this merits its own dedicated slide.
> 
> I think we do briefly when we say we are making a new model.

Can you point me to the location so I can take a look?

**Comment by mgelbart** [comment]
Ex 8 looks good. Need to replace "medium" -> "median".

**Comment by hfboyce** [comment]
> In the current version, we call train_test_split on an earlier slide. Then here we glue them back together and then scale to show that this is a bad idea. Instead, why not just take the original df from way back before train_test_split was ever called?

We do this because we teach them to impute separately so, we carried on from the separately imputed data. 

>Not sure if I'm seeing the new version or if the refresh failed. Can you send me a screenshot to confirm?

![image](https://user-images.githubusercontent.com/42727648/98070101-58ecea00-1e15-11eb-8443-db27060362bd.png)
![image](https://user-images.githubusercontent.com/42727648/98070117-60ac8e80-1e15-11eb-9aa1-8c531920fa0c.png)
![image](https://user-images.githubusercontent.com/42727648/98070128-64d8ac00-1e15-11eb-815b-b839f23149d9.png)

> 20: I got the first MCQ wrong, whoops! But, do we mention the refitting anywhere in the slides? I think this merits its own dedicated slide.
I think we do briefly when we say we are making a new model.
Can you point me to the location so I can take a look?

Slides 17.12 and 17.13 I amended since you were right it wasn't in there and I fixed up Question 20 for efficiency.

**Comment by mgelbart** [comment]
1. OK, seems like we don't agree. I will back down on this one.
2. Yeah, I'm still not seeing the new version. But the images you posted look great.
3. Again, I'm not seeing it. I guess I should clear my cache - will do that later. But sounds good.
